patches/server/0962-Registry-and-Lifecycle-API.patch

```diff
@@ -0,0 +1,1926 @@
++                this.helpSchedule(pair.lifecyclePoint(), pair.consumer(), pluginMeta);
++            }
++        }
++    }
++
++    @Override
++    public boolean isBuilt() {
++        return built.get();
++    }
++
++    record Pair<T>(LifecyclePoint<T> lifecyclePoint, Consumer<T> consumer) {}
++
++    @SuppressWarnings("unchecked")
++    private <T> void helpSchedule(final LifecyclePoint<?> lifecyclePoint, final Consumer<?> consumer, final PluginMeta pluginMeta) {
++        ((LifecyclePoint<T>) lifecyclePoint).schedule((Consumer<T>) consumer, pluginMeta);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/lifecycle/LifecyclePointUtil.java b/src/main/java/io/papermc/paper/lifecycle/LifecyclePointUtil.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..baef0b51dac3cb407816ddfef2d8b3b99845e6c4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/lifecycle/LifecyclePointUtil.java
+@@ -0,0 +1,91 @@
++package io.papermc.paper.lifecycle;
++
++import io.papermc.paper.registry.APIConverters;
++import io.papermc.paper.registry.Converters;
++import io.papermc.paper.registry.ExtendedRegistry;
++import io.papermc.paper.registry.ExtendedRegistryImpl;
++import io.papermc.paper.registry.RegistryAPI;
++import io.papermc.paper.registry.key.RegistryKey;
++import io.papermc.paper.registry.key.ResourceKey;
++import net.kyori.adventure.key.Key;
++import net.minecraft.core.Registry;
++import net.minecraft.resources.ResourceLocation;
++import org.bukkit.Keyed;
++
++import java.util.Collection;
++import java.util.HashMap;
++import java.util.Map;
++import java.util.Optional;
++
++public final class LifecyclePointUtil {
++    public static final SimpleRegistryImpl SIMPLE_REGISTRY = new SimpleRegistryImpl();
++
++    private LifecyclePointUtil() {}
++
++    public static <C> void enterRegistryInitialization(LifecyclePoint<C> lifecyclePoint, C context, Collection<? extends Registry<?>> registries) {
++        SIMPLE_REGISTRY.registryMap.put(lifecyclePoint, registries);
++        Optional.ofNullable(SIMPLE_REGISTRY.actionMap.get(lifecyclePoint)).ifPresent(Runnable::run);
++        LifecyclePoints.enter(lifecyclePoint, context);
++    }
++
++    public static class SimpleRegistryImpl implements ExtendedRegistry.SimpleRegistry {
++        final Converters.CachedConverter<Key, ResourceLocation> keyConverter = Converters.KEY_RESOURCE_LOCATION_CONVERTER.get();
++        final Map<Class<? extends RegistryKey.RegistryLayerType>, LifecyclePoint<?>> map = new HashMap<>();
++        final Map<LifecyclePoint<?>, Collection<? extends Registry<?>>> registryMap = new HashMap<>();
++        final Map<LifecyclePoint<?>, Runnable> actionMap = new HashMap<>();
++
++        {
++            map.put(RegistryKey.RegistryLayerType.STATIC.class, ServerLifecyclePoints.STATIC_REGISTRIES_INITIALIZED);
++            map.put(RegistryKey.RegistryLayerType.WORLDGEN.class, ServerLifecyclePoints.WORLDGEN_REGISTRIES_INITIALIZED);
++        }
```

Comment:
I guess i am trying to wrap my head around these registry layer types. Specifically, why do these types need to be denoted separately?

Reply:
These types are used in RegistryKey in a generic as a compile-time check, so users in WritableRegistryAccess can get WritableRegistry only if the RegistryKey's LayerType generic matches the generic of WritableRegistryAccess, so people won't misuse lifecycle api, i.e. try to get a worldgen registry from the static registry access.